concurrent/src/main/java/io/airlift/concurrent/AsyncQueue.java

```diff
@@ -0,0 +1,118 @@
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A simple unbounded queue that can fetch batches of elements asynchronously.
+ *
+ * This class is designed for multiple writers but only a single reader, since
+ * as single shared future is shared for all readers.
+ */
+public class AsyncQueue<T>
+{
+    private enum FinishedMarker
+    {
+        FINISHED
+    }
+
+    private final BlockingQueue<Object> queue = new LinkedBlockingQueue<>();
+    private final AtomicReference<CompletableFuture<?>> futureReference = new AtomicReference<>(new CompletableFuture<>());
+
+    /**
+     * Adds an element to the queue.  Elements added after the queue is finished
+     * are ignored without error.
+     * @throws NullPointerException if the element is null
+     */
+    public void add(T element)
+    {
+        requireNonNull(element, "element is null");
+        queue.add(element);
+        futureReference.get().complete(null);
+    }
+
+    /**
+     * Gets a batch of elements from the queue.  If queue is not empty, a
+     * completed future containing the queued elements is returned.  If the
+     * queue is empty, an incomplete future is returned that completes when
+     * an element is added.
+     *
+     * Note: this class is designed for multiple writers but only a single
+     * reader, since as single shared future is shared for all readers.
+     *
+     * @param maxSize the maximum number of elements to return.
+     */
+    public CompletableFuture<List<T>> getBatchAsync(int maxSize)
+    {
+        return futureReference.get().thenApply(x -> getBatch(maxSize));
+    }
+
+    private List<T> getBatch(int maxSize)
+    {
+        // wait for at least one element and then take as may extra elements as possible
+        // if an error has been registered, the take will succeed immediately because
+        // will be at least one finished marker in the queue
+        List<Object> elements = new ArrayList<>(maxSize);
+        queue.drainTo(elements, maxSize);
+
+        // check if we got the finished marker in our list
+        int finishedIndex = elements.indexOf(FinishedMarker.FINISHED);
+        if (finishedIndex >= 0) {
+            // add the finish marker back to the queue so future callers will not block indefinitely
+            queue.add(FinishedMarker.FINISHED);
+            // drop all elements after the finish marker (this shouldn't happen in a normal finish, but be safe)
+            elements = elements.subList(0, finishedIndex);
+        }
+
+        // if the queue is empty and the current future is finished, create a new one so
+        // readers can be notified when the queue has elements to read
+        if (queue.isEmpty()) {
```

Comment:
there is a race condition here:
1) reader checks queue.isEmpty() => true
2) writer adds element to the queue
3) writer sets future to complete
4) reader swaps the old future to the new future since it the old is complete
5) reader waits on the batch, but never realizes that it already has data available

The same thing can happen if this races with the finish() call, the readers can miss the FINISH signal.